apps/aeoracle/src/aeo_query_tx.erl

```diff
@@ -0,0 +1,202 @@
+
+-spec query_ttl(query_tx()) -> aeo_oracles:ttl().
+query_ttl(#oracle_query_tx{query_ttl = QueryTTL}) ->
+    QueryTTL.
+
+-spec response_ttl(query_tx()) -> aeo_oracles:relative_ttl().
+response_ttl(#oracle_query_tx{response_ttl = ResponseTTL}) ->
+    ResponseTTL.
+
+-spec new(map()) -> {ok, query_tx()}.
+new(#{sender        := SenderPubKey,
+      nonce         := Nonce,
+      oracle        := Oracle,
+      query         := Query,
+      query_fee     := QueryFee,
+      query_ttl     := QueryTTL,
+      response_ttl  := ResponseTTL,
+      fee           := Fee}) ->
+    {ok, #oracle_query_tx{sender        = SenderPubKey,
+                          nonce         = Nonce,
+                          oracle        = Oracle,
+                          query         = Query,
+                          query_fee     = QueryFee,
+                          query_ttl     = QueryTTL,
+                          response_ttl  = ResponseTTL,
+                          fee           = Fee}}.
+
+-spec fee(query_tx()) -> integer().
+fee(#oracle_query_tx{fee = F}) ->
+    F.
+
+-spec nonce(query_tx()) -> non_neg_integer().
+nonce(#oracle_query_tx{nonce = Nonce}) ->
+    Nonce.
+
+-spec origin(query_tx()) -> pubkey().
+origin(#oracle_query_tx{sender = SenderPubKey}) ->
+    SenderPubKey.
+
+%% SenderAccount should exist, and have enough funds for the fee + the query_fee.
+%% Oracle should exist, and query_fee should be enough
+%% Fee should cover TTL
+%% TODO: Check the query? Check TTL relative to Oracle TTL?
+-spec check(query_tx(), trees(), height()) -> {ok, trees()} | {error, term()}.
+check(#oracle_query_tx{sender = SenderPubKey, nonce = Nonce,
+                       oracle = OraclePubKey, query_fee = QFee,
+                       query_ttl = TTL, fee = Fee}, Trees, Height) ->
+    Checks =
+        [fun() -> aetx_utils:check_account(SenderPubKey, Trees, Height, Nonce, Fee + QFee) end,
+         fun() -> aeo_utils:check_ttl_fee(Height, TTL, Fee - ?ORACLE_QUERY_TX_FEE) end,
+         fun() -> check_oracle(OraclePubKey, Trees, QFee) end],
+
+    case aeu_validation:run(Checks) of
+        ok              -> {ok, Trees};
+        {error, Reason} -> {error, Reason}
+    end.
+
+-spec signers(query_tx()) -> [pubkey()].
+signers(#oracle_query_tx{sender = SenderPubKey}) ->
+    [SenderPubKey].
+
+-spec process(query_tx(), trees(), height()) -> {ok, trees()}.
+process(#oracle_query_tx{sender = SenderPubKey, nonce = Nonce, fee = Fee,
+                         query_fee = QueryFee} = Query, Trees0, Height) ->
+    AccountsTree0 = aec_trees:accounts(Trees0),
+    OraclesTree0  = aec_trees:oracles(Trees0),
+
+    Sender0 = aec_accounts_trees:get(SenderPubKey, AccountsTree0),
+    {ok, Sender1} = aec_accounts:spend(Sender0, QueryFee + Fee, Nonce, Height),
+    AccountsTree1 = aec_accounts_trees:enter(Sender1, AccountsTree0),
+
+    Interaction = aeo_interaction:new(Query, Height),
+    OraclesTree1 = aeo_state_tree:enter_interaction(Interaction, OraclesTree0),
```

Comment:
I understand assumption is that interaction is not there. I suggest prefering `aeo_state_tree:insert_interaction` to `aeo_state_tree:enter_interaction` following gb_trees API semantics wrt crash (nice to have, for clarity).

Reply:
Will fix

Reply:
Added insert_interaction as suggested.